src/Microsoft.AspNetCore.Server.Kestrel/Internal/Infrastructure/MemoryPoolIterator.cs

```diff
@@ -1059,5 +1058,174 @@ public unsafe void CopyFromAscii(string data)
             _block = block;
             _index = blockIndex;
         }
+
+        public unsafe string GetAsciiString(ref MemoryPoolIterator end)
+        {
+            if (IsDefault || end.IsDefault)
+            {
+                return null;
+            }
+
+            var length = GetLength(end);
+
+            if (length == 0)
+            {
+                return null;
+            }
+
+            var inputOffset = Index;
+            var block = Block;
+
+            var asciiString = new string('\0', length);
+
+            fixed (char* outputStart = asciiString)
+            {
+                var output = outputStart;
+                var remaining = length;
+
+                var endBlock = end.Block;
+                var endIndex = end.Index;
+
+                var outputOffset = 0;
+                while (true)
+                {
+                    int following = (block != endBlock ? block.End : endIndex) - inputOffset;
+
+                    if (following > 0)
+                    {
+                        if (!AsciiUtilities.TryGetAsciiString(block.DataFixedPtr + inputOffset, output + outputOffset, following))
+                        {
+                            throw BadHttpRequestException.GetException(RequestRejectionReason.NonAsciiOrNullCharactersInInputString);
+                        }
+
+                        outputOffset += following;
+                        remaining -= following;
+                    }
+
+                    if (remaining == 0)
+                    {
+                        break;
+                    }
+
+                    block = block.Next;
+                    inputOffset = block.Start;
+                }
+            }
+
+            return asciiString;
+        }
+
+        public string GetUtf8String(ref MemoryPoolIterator end)
+        {
+            if (IsDefault || end.IsDefault)
+            {
+                return default(string);
+            }
+            if (end.Block == Block)
+            {
+                return Constants.UTF8.GetString(Block.Array, Index, end.Index - Index);
```

Comment:
It seems really weird to me to prefer `Constants.UTF8` over `Encoding.UTF8`. Just glancing at this code I would assume that Kestrel used some custom encoder other than `Encoding.UTF8`.